### ISSUE TITLE
Support more complex CI / test suite / partial coverage setups

### ISSUE THREAD

**Issue by pcardune** [codecov/support#355]
At https://github.com/code-dot-org/code-dot-org, we have _a lot_ of tests that take _a very long time_ to run. To optimize our continuous integration setup, we've had to depart from the typical behavior of smaller projects.

Some background about our codebase to guide discussion:
* we have a [mono repo](http://danluu.com/monorepo/) that is comprised of several sub projects which interface with each other in complicated ways. The projects are the following:
  * `apps` - a large javascript code base with both unit and integration tests
  * `pegasus` - a large ruby/sinatra code base with it's own set of unit tests.
  * `dashboard` - a large ruby on rails code base with both unit tests and end-to-end browser tests. The end-to-end browser tests exercise code in `dashboard`, `apps`, _and_ `pegasus`.
* We use CircleCI for continuous integration. Each CI run takes about an hour, running on two parallel containers. Each container runs a different set of tests. At any given time during the day, we have many CI runs going in parallel.
* Each CI run determines which test suite to run based on which files have changed:
  * If no application code was changed (i.e. someone just updated a README file), then no test suites are run.
  * If only files in the `apps` code base were changed, then we only run unit/integration tests and end-to-end browser tests, but skip the `dashboard` and `pegasus` unit tests.
  * If only files in the `dashboard` code base were changed, then we only run the `dashboard` unit tests and the end-to-end browser tests.
* CI is run for every pull request, and for every commit to our main branch (called `staging`)
* for CI runs triggered from a pull request, the tests are run against the PR branch HEAD commit, _**without**_ merging with the main `staging` branch first.
* We _rarely_ commit directly to the main `staging` branch. Instead, we almost always open a pull request, and then use a merge commit to get the code onto the `staging` branch.

Since each commit can have a somewhat arbitrary set of tests run against it, comparing coverage reports between commits isn't particularly useful, and is actually really confusing to look at. However, using Codecov's [flag feature](https://docs.codecov.io/docs/flags), there are opportunities to greatly increase codecov's usefulness for such a complex code base.

For example, suppose we have the following series of (merge) commits, with `x` marking which test suite was run. Coverage reports are generated for all test suites except browser-e2e:

| Commit | description | apps-unit | apps-integration | dashboard-unit | pegasus-unit | browser-e2e |
|---|---|---|---|---|---|---|
| A | wrote some js code in apps | x | x | - | - | x |
| B | wrote some ruby in dashboard | - | - | x | - | x |
| C | updated a README file | - | - | - | - | - |
| D | wrote lots of code everywhere | x | x | x | x | x |
| E | wrote some ruby in pegasus | - | - | - | x | x |

**Feature 1: merge according to flag across multiple commits**
What I would like to see on the codecov dashboard is the merging of all coverage across the different test suites, using whatever coverage information is available.

| Commit | coverage merge calculation |
|---|---|
| A | apps-unit@A + apps-integration@A |
| B | apps-unit@A + apps-integration@A + dashboard-unit@B |
| C | apps-unit@A + apps-integration@A + dashboard-unit@B |
| D | apps-unit@D + apps-integration@D + dashboard-unit@D + pegasus-unit@D |
| E | apps-unit@D + apps-integration@D + dashboard-unit@D + pegasus-unit@E |

So for each commit, in order to calculate the overall coverage report, you merge up coverage reports for each flag from the last commit for which coverage info with that flag was available.

Or stated differently, for any given flag, if coverage information is not available for a particular commit, keep going up the commit history until you find coverage information for that flag, and then assume that the coverage has not changed since then.

You would then use these merged coverage reports to do per commit coverage diffing.

**Feature 2: coverage diff between PR HEAD and nearest common ancestor to PR BASE**
When computing coverage diffs for pull requests that look like this:
```
master . A . . B . . C . . D . . E
                \
pull              . . F . . G
```
instead of reporting a coverage diff between `E` and `G`, which would be missing any coverage that got added in `C . . D . . E`, you would report a coverage diff between `B` and `G`.

**Comment by stevepeak** [comment]
Thank you very much for this detailed issue. I'll be reviewing this in more detail and certainly follow up with questions. But here are some quick notes:

## Feature 1

This is an interesting concept. The logic seams sounds, however based on how the system works this may require some major refactoring of the code base.
We are planning some refactoring to grow the feature set and can keep this in mind to see if it is possible.

## Feature 2

> instead of reporting a coverage diff between E and G

Codecov would **never** do this. It will always compare `B...G` (B is the pr base)

Related to of #345 and #354

**Comment by stevepeak** [comment]
@pcardune first step would be flagging these uploads http://docs.codecov.io/docs/flags

I would like to see if adding Flags changes your approach to this issue. Thank you!

**Comment by stevepeak** [comment]
```yaml
flags:
  python:
    # if python flag not provided then assume the parent commits coverage
    assumed: true
  smoketests:
    # smoketests will always run on all commits
    assumed: false  # assumed
```

## Description
```
when commit completed uploading
  for each missing flag
    if flag is ASSUMED and was not uploaded
      use parent commit's coverage metrics for flag[name]
```

## Keep in mind
- respect `flags.<>.paths`
- flags may not upload and use global data #366
 

-----

@pcardune could you mock up your yaml using assumed metrics?

**Comment by BBonifield** [comment]
The above proposed solution sounds perfect to me.  The only additional bit of configuration that might be useful would be to have the `assumed` flag have whitelist or blacklist branches.  For instance, you might only want to assume coverage on PRs whereas you would expect all `master` builds to provide full coverage.  That being said, that configurability isn't important to us.  It'd just be a nice to have.

**Comment by stevepeak** [comment]
@BBonifield Very nice feedback 👍 If you do not mind: provide an example yaml with this `assumption` concept.

Your suggested changes would look like.
```yaml
flags:
  python:
    # if python flag not provided then assume the parent commits coverage
    assume:
      branches:  # all branches are assumed if not provided
        - master  # only assume #python when parent commit is on master branch
  smoketests:
    # smoketests will always run on all commits
    assume: false
```

**Comment by BBonifield** [comment]
@stevepeak I actually think the inverse might be the more useful one:

```
flags:
  python:
    assume:
      except_branches:
        - master # assume parent coverage always except if the SHA is a commit on master
```

I imagine it'd work like:

```
9aa44c8 # commit on master - branch created from here
  \- 38a6b22 # first commit on branch, assumes 9aa44c8 coverage for python
   |- 3a836ca # second commit on branch, assumes 9aa44c8 coverage for python
   |- 826bc26 # third commit on branch, assumes 9aa44c8 coverage for python
  /
64a84b3 # merge commit into master - requires coverage
```

**Comment by stevepeak** [comment]
@BBonifield Awesome feedback!

`except_branches=master` would be equivalent to `branches=!master` 

```yaml
branches:
  - !master
```
> considering other codecov.yml configuration uses `branches`.

Thoughts?

**Comment by BBonifield** [comment]
@stevepeak Makes perfect sense to me!

**Comment by stevepeak** [comment]
I'll be working on this during the weekend. Any other feedback before I dive into it? Meet me in [Slack](slack.codecov.io) to test and beta. Thank you!

**Comment by TimvdLippe** [comment]
Got redirected here from #345. We are facing this issue with Mockito, as we ci skip our release commits. Codecov is therefore unable to generate a report and needs to use its parent. Any ETA for this feature?

**Comment by interstateone** [comment]
My initial hopes for Codecov based on the description of the flags and merging features were that it would work for the project I'm working on right now, which has a similar setup to the one in this issue.

I assumed that when only certain flags were uploaded that the previous coverage values would be used. In practice, the coverage values in the PR comments and on the dashboard that are made by Codecov when only certain flags are uploaded seem really bizarre and are often conflicting (not sure if this is a bug or an artifact of our CI uploading coverage data in a weird way).

I'm sure there's a reason that coverage is calculated the way that it is, but it's not useful to us as-is. There must be a use case that would set `assume: false` (or for the current implementation to default to this) but I'm curious what it is... If CI succeeds and Codecov succeeds when not all flags are uploaded but _some_ are, I would think that `assumes: true` would be a better default behaviour? Again, I'm probably ignoring another use case.

In any case, the initial suggestion to add an `assume` property to a flag would fit our needs well. I'm not sure we need to filter based on branches but that wouldn't hurt us either.

**Comment by stevepeak** [comment]
@TimvdLippe Your issue is not quite related to this. Can you ping me in [Slack](https://slack.codecov.io) or [Gitter](https://gitter.im/codecov/support) to review your project. Thanks!

@interstateone ::

When you only upload certain flags then the comment would be off because it's missing data. The concept of "assuming" data is not yet implemented.

Given that "assuming" is not implemented: The default behavior is `assume: false` this is because for the vast majority of projects they upload all flags on every build. If default was `assume: true` then status/comments would appear successful regardless of missing data. We can have a discussion concerning the default behavior once it is implemented.

This feature is on our radar. It's really cool+useful and we look forward to implementing it. There are technical challenges to overcome that we are reviewing: e.g. some reports are so large that if assumptions are used it could require pulling several reports at once and significantly impact performance of the product.

**Comment by TimvdLippe** [comment]
@stevepeak I have contacted you on Gitter

**Comment by ileitch** [comment]
> I'll be working on this during the weekend. Any other feedback before I dive into it? Meet me in Slack to test and beta. Thank you!

@stevepeak any progress on this?

**Comment by stevepeak** [comment]
It's still on the radar - but requires quite a lot of work to pull off.

**Comment by Blackbaud-PatrickOFriel** [comment]
@stevepeak @TimvdLippe Did you ever resolve the issue mentioned above. We're running into some wonkiness when we have a ci skip when committing baseline screenshots to our repository

**Comment by janpaul123** [comment]
So this is how we have it set up now:

<img width="412" alt="screen shot 2017-08-18 at 17 35 48" src="https://user-images.githubusercontent.com/177461/29481879-ba416f3c-843b-11e7-8c91-db0a587eeda4.png">

For now we're going to set the threshold to 0% for all of them, but a "ignore if coverage is 0%" flag would basically solve this for us! 😄

**Comment by stevepeak** [comment]
@janpaul123 You'll need to set the `target: 0%` to accomplish what you mention above.

**Comment by stevepeak** [comment]
CC @Blackbaud-PatrickOFriel 

**General Update** This issue has been requested more often now. As you can understand, it is a complex issue that requires the system to pull multiple reports during a compare. We are working on a solution for this by re-architecturing our reporting classes. Obviously, taking longer than expected due to rapid company growth.

This issue remains on the close horizon. We really appreciate your patience and understanding. 

> Keep in mind that this feature will require the usage of [Flags](https://docs.codecov.io/docs/flags). Flags will be at the center of Codecov features moving forward.

**Comment by aayers-agrible** [comment]
This is really great! I have a CI setup that involves only the applications which have changed getting build and tested in iOS. They all upload their test results under different flags, and what I really want to know is the change in coverage for a given flag vs. the last set of results for that flag. If I understand this correctly, the proposed feature would solve my issue entirely.

**Comment by stevepeak** [comment]
@aayers-agrible 💯 This is exactly what this feature will solve.

**Comment by stevepeak** [comment]
## Statuses for dynamic Flags

> Thanks to @mattclay 👏 👍 

Project has N flags (maybe hundreds in the future).
Developer would like to have a commit status that concerns only the flag(s) that are uploaded during a CI build. The following (*concept*) yaml would be an example of how to use **Dynamic Statuses**.

```yaml
coverage:
  status:
    project:
      foobar:  # this is a custom title
        only_pulls: yes  # only post this status on pull requests (optional)
        flags:
          - "test-*"  # using glob/regexp to match flags
        style: single  # choose: _single_ or _many_ flags per combined in the same status
```

### Example Pull 1
- Uploads only one flag: `test-a`.
- Gets one status with context: `codecov/project/foobar/test-a`

## Example Pull 2
- Uploads two flags: `test-nb` and `test-c`.
- Gets two statuses with context: `codecov/project/foobar/test-a` and `codecov/project/foobar/test-b`

-----

Feedback very, very welcome.

**Comment by aayers-agrible** [comment]
One question I would have about this feature is how the overall project coverage would be resolved, for instance in the graph that is shown on the main page of the project. What number would be used to calculate the "min per day"? Would it be from the individual statuses that were posted that day, or could it be the overall resolved coverage from the most recent version of every flag up to that point?

**Comment by mrcasals** [comment]
Hi all! 

I think I'm having the same issue as @pcardune, but I want to check it. I'm one of the maintainers of https://github.com/decidim/decidim, as you can see we are using a monorepo strategy right now. Most of our changes inly affect one of the folders, so I wanted to run the tests for only the folders that have modified files. This is the PR implementing this change:

https://github.com/decidim/decidim/pull/2028

We're using CircleCI 2.0 with workflows, with one job per folder. Last commit in the PR (`6a676af Remove test file`) correctly runs the tests for one of the test jobs, while the others are skipped:

https://circleci.com/workflow-run/ea5c9842-4426-42ca-bc49-4e443c130335 (onnly `main` is running the tests)

Unfortunately, codecov is waiting for the status check. I guess codecov is expecting to receive the rest of the coverage, but I'd like it to only consider the lines changed, and consider the other folders as untouched and keep the coverage reports. Is this issue related to mine? Should I open a new issue?

I'm trying the flags feature here, but I'm not having much success: 

https://github.com/decidim/decidim/pull/2045

any help would be welcome 🙏  Thanks in advance!

**Comment by stevepeak** [comment]
@mrcasals Codecov will wait for all Commit Statues, filtered to only CI testing related, to be reported. This is key to posting accurate Codecov commit statuses.

In your case: there are many Circle statuses, which is perfectly fine. I see the latest commit is working fine, so maybe this a non-issue. But there is some complexity behind our checks, but for your case we rely on GitHub webhooks which get pushed with Circle update the commit status telling Codecov to check if all of the CI is complete or not. This process appears to have worked for your commit decidim/decidim@6f69e61

Please let me know if you have more questions. If you run *all* your tests during each Circle build then this specific thread/issue does not apply to you.

**Comment by mrcasals** [comment]
@stevepeak we just merged the PR adding flags, and I rebased https://github.com/decidim/decidim/pull/2028 so it's using flags. To show the code is working, I've added a new file in the `decidim-proposals/` folder, so only the `proposals` and `main` jobs would run the tests, and thus report status to codecov.

This is the codecov report:

https://codecov.io/gh/decidim/decidim/pull/2028

As you can see, there's a lot of changes in the report. In the files list, the only folder that has no tracked lines change is `proposals`, which is where the tests run. What I want to achieve is that the report does not show any modification for other folders if tests did not run there. Is this possible to achieve with flags?

This is my current `codecov.yml` file for that PR branch: https://github.com/decidim/decidim/blob/tests/conditionally-pass-tests/codecov.yml

**Comment by stevepeak** [comment]
We will be **deploying this feature shortly**. I wanted to let you all know what will be supported in the first run. 

## ✅  Added `assume` to flags
```yaml
flags:
  sub_project_a:
    assume: true  # if not uploaded it will be assumed after commit finishes CI
  sub_project_b:
    assume: 
      branches:
        - master  # only assume when on the master branch
```
> Default is **not** to assume.

#### How assumptions work
1. You upload report(s) to Codecov.
1. We wait for CI to completes.
1. Once CI is complete, we loop through all flags to see if any are missing and should be assumed.
1. We inherit coverage from the previous parent report that should be assumed.

## 🔜  Dynamic Statuses
> Coming soon. The concept is outlined here https://github.com/codecov/support/issues/355#issuecomment-326334500

----

### Join the conversation at https://slack.codecov.io in the group `#assuming` for all QA/Support. Thanks
